[Feature] 태그 등록문제 해결 및 리뷰 조회 구현

ghrltjdtprbs opened the pull request:
## ⭐[feature] 리뷰 조회 구현

### [전체 리뷰 조회]
- 등록되어 있는 전체 리뷰를 조회
- 최신순(default), 별점순 정렬 가능
- 로그인 없이도 조회 가능
- 부트캠프 파라미터 없을시 전체 조회
   -> ex ) GET /api/v2/reviews/all

### [부트캠프별 리뷰 조회]
- 부트캠프별 리뷰를 조회
- 최신순(default), 별점순 정렬 가능
- 로그인 없이도 조회 가능
   -> ex )  GET /api/v2/reviews?bootcamp=부트캠프 이름
-------------------------
### [부트캠프 Summary]
- 부트캠프별 리뷰 정보를 요약하는 api 구현
- 부트캠프 이름, 평균 별점, 리뷰의 개수를 출력합니다.
- 로그인 없이도 조회 가능
- 엔드포인트 : GET /api/v2/reviews/summary
------------------------
### [부트캠프별 태그 그래프]
- 부트캠프별 태그 그래프 구현을 위한 api 구현
- 부트캠프별 총 사용 태그갯수, 개별 태그 개수를 출력합니다.
- 로그인 없이도 조회 가능
- 엔드포인트 : GET /api/v2/reviews/tag-graph?bootcamp=부트캠프 이름
-------------------
## ⭐[fix] 리뷰 등록, 수정 시 태그 문제 해결

문제 사항 :
 - 리뷰를 등록, 수정시 review_tag 테이블에 데이터 저장이 안됨.
 - goodtag, badtag 구분을 안하고 저장함 -> 조회시 태그 구분 불가 


변경 사항 :
 - Review 엔터티에 tag cascade 추가
 - goodtag, badtag 구분을 위해 ReviewTag에 boolean타입의 isGoodTag 추가
 - 리뷰 수정, 재등록시 기존 리뷰 태그 삭제 후 재등록하는 것으로 로직 수정

ghrltjdtprbs committed:
[fix] 리뷰 등록, 수정시 태그 등록문제 해결

문제 사항 :
 - 리뷰를 등록, 수정시 review_tag 테이블에 데이터 저장이 안됨.
 - goodtag, badtag 구분을 안하고 저장함 -> 조회시 태그 구분 불가
변경 사항 :
 - Review 엔터티에 tag cascade 추가
 - goodtag, badtag 구분을 위해 ReviewTag에 boolean타입의 isGoodTag 추가
 - 리뷰 수정, 재등록시 기존 리뷰 태그 삭제 후 재등록하는 것으로 로직 수정

ghrltjdtprbs committed:
[feature] 전체리뷰 조회 구현

변경 사항 :
 - 등록되어 있는 전체 리뷰를 조회하는 api 구현
 - 최신순(default), 별점순 정렬 가능
 - 로그인 없이도 조회 가능
 - 엔드포인트 : GET /api/v2/reviews/all

ghrltjdtprbs committed:
[feature] 부트캠프별 리뷰 조회 구현

변경 사항 :
 - 부트캠프별 리뷰를 조회하는 api 구현
 - 최신순(default), 별점순 정렬 가능
 - 로그인 없이도 조회 가능
 - 엔드포인트 : GET /api/v2/reviews/by-bootcamp

ghrltjdtprbs committed:
[feature] 부트캠프별 리뷰 요약 구현

변경 사항 :
 - 부트캠프별 리뷰를 요약하는 api 구현
 - 부트캠프 이름, 평균 별점, 태그 총 개수, 태그별 개수, 리뷰의 개수를 출력합니다.
 - 로그인 없이도 조회 가능
 - 엔드포인트 : GET /api/v2/reviews/by-bootcamp/summary

ghrltjdtprbs committed:
[feature] 리뷰작성 restdocs 테스트코드 구현

변경 사항 :
 - 리뷰작성 restdocs 테스트코드를 구현하였습니다

ghrltjdtprbs committed:
[feature] 리뷰삭제 restdocs 테스트코드 구현

변경 사항 :
 - 리뷰삭제 restdocs 테스트코드를 구현하였습니다

ghrltjdtprbs committed:
[feature] 리뷰수정 restdocs 테스트코드 구현

변경 사항 :
 - 리뷰수정 restdocs 테스트코드를 구현하였습니다
 - 기존 테스트코드 응답값 수정

ghrltjdtprbs committed:
[feature] 전체 리뷰조회 restdocs 테스트코드 구현

변경 사항 :
 - 전체 리뷰조회 restdocs 테스트코드를 구현하였습니다
 - queryparam 임포트 에러 문제로 제외하고 구현

ghrltjdtprbs committed:
[feature] 부트캠프별 리뷰조회 restdocs 테스트코드 구현

변경 사항 :
 - 부트캠프별 리뷰조회 restdocs 테스트코드를 구현하였습니다
 - queryparam 임포트 에러 문제로 제외하고 구현

ghrltjdtprbs committed:
[feature] 부트캠프별 리뷰요약 restdocs 테스트코드 구현

변경 사항 :
 - 부트캠프별 리뷰요약 restdocs 테스트코드를 구현하였습니다

ghrltjdtprbs committed:
[feature] review-api.doc 작성

변경 사항 :
 - review-api.doc 작성

ghrltjdtprbs committed:
[feature] ReviewControllerTest 작성

변경 사항 :
 - ReviewControllerTest 작성

ghrltjdtprbs committed:
[refactor] ReviewControllerTest 메소드명 변경

변경 사항 :
 - convention에 맞게 메소드명을 변경하였습니다.

ghrltjdtprbs committed:
[feat] Review엔터티 테스트 작성

변경 사항 :
 - Review엔터티 테스트를 작성하였습니다.

ghrltjdtprbs committed:
[feat] Tag엔터티 테스트 작성

변경 사항 :
 - Tag 엔터티 테스트를 작성하였습니다.

ghrltjdtprbs committed:
[feat] ReviewServiceTest 작성

변경 사항 :
 - ReviewServiceTest 테스트를 작성하였습니다.

Kongji82 commented on src/main/java/com/fasttime/domain/review/entity/ReviewTag.java:
```diff
@@ -28,4 +28,6 @@ public class ReviewTag {
     @JoinColumn(name = "tag_id")
     private Tag tag;
 
+    private boolean isGoodTag;
```
good tag, bad tag 구분은 Review가 아니라 Tag 엔티티에서 하는거 어떻게 생각하시나요?

ghrltjdtprbs replied:
구현 중에 임의로 추가한 부분이라 중간테이블에 isGoodTag필드를 넣었습니다! 말씀대로 Tag 엔터티에서 처리하는게 좋을 것 같네요 ㅎㅎ 수정하겠습니다! 😊😊

Kongji82 commented on src/main/java/com/fasttime/domain/review/service/ReviewService.java:
```diff
@@ -28,38 +29,32 @@
         if (existingReview != null) {
             if (existingReview.isDeleted()) {
+                reviewTagRepository.deleteByReview(existingReview);
+                updateReviewDetails(existingReview, requestDTO);
+                updateReviewTags(existingReview, requestDTO);
```
중복 부분이 있어서 따로 함수화해도 될것 같습니다.

ghrltjdtprbs replied:
피드백 반영해 수정하였습니다! 44c10b8209d695ae2ddf631f3813ff769be958e3 커밋 확인해주세요 👍👍

Kongji82 commented on src/main/java/com/fasttime/domain/review/service/ReviewService.java:
```diff
@@ -83,30 +78,48 @@ private Set<String> getTagContents(Set<Long> tagIds) {
         }
+        reviewTagRepository.deleteByReview(review);
 
+        updateReviewDetails(review, requestDTO);
+        updateReviewTags(review, requestDTO);
```
(중복)

JeongUijeong commented on src/main/java/com/fasttime/domain/review/controller/ReviewController.java:
```diff
@@ -54,4 +58,26 @@ public ResponseEntity<ResponseDTO<ReviewResponseDTO>> updateReview(
             ResponseDTO.res(HttpStatus.OK, REVIEW_SUCCESS_MESSAGE, responseDTO));
     }
+
+    @GetMapping("/all")
```
RESTful API의 조회는 collection 목록을 받는 행위입니다! 따라서 `/all`을 PATH에 붙일 필요가 없습니다~
PATH에 붙이는 조건은 필터라고 생각하시고 설계하시면 좋다고 합니다! 😊

JeongUijeong replied:
추가적으로 부트캠프 후기 목록 조회에서 페이지네이션 기능이 필요 없을까요?

ghrltjdtprbs replied:
/all을 붙인 이유는 securityconfig에 등록하기 위해 임의로 붙였습니다! 페이지네이션은 저도 생각은 했으나, 프론트와 상의 후에 붙이려고 일단 보류해뒀습니다!

JeongUijeong replied:
넵 답변 감사합니다! 😊
```
.requestMatchers(HttpMethod.GET, "api/v2/reviews").permitAll()
```
securityconfig에서는 위처럼 GET 매핑에 대해서만 허용해두시면 됩니다~

ghrltjdtprbs replied:
피드백 반영해 엔드포인트 수정했습니다! eaa7e659c2555b5b3bbec6c2650705021384784a 커밋 확인해주세요 😊😊

JeongUijeong commented on src/main/java/com/fasttime/domain/review/controller/ReviewController.java:
```diff
@@ -54,4 +58,26 @@ public ResponseEntity<ResponseDTO<ReviewResponseDTO>> updateReview(
+        return ResponseEntity.ok(ResponseDTO.res(HttpStatus.OK, REVIEW_SUCCESS_MESSAGE, reviews));
+    }
+
+    @GetMapping("/by-bootcamp")
+    public ResponseEntity<ResponseDTO<List<ReviewResponseDTO>>> getReviewsByBootcamp(
+        @RequestParam String bootcamp,
+        @RequestParam(required = false, defaultValue = "createdAt") String sortBy) {
```
getReviews와 다른 점이 쿼리 파라미터로 bootcamp 문자열을 받는 것 뿐인 것 같습니다~ 
getReviews에서 쿼리 파라미터로 bootcamp 문자열을 받게 한 후 required = false로 설정하시는 방법은 어떤가요?
그 후 Service나 Repository에서 처리하시면 좋을 것 같습니다. 
1. 쿼리 파라미터로 bootcamp에 문자열을 담아 요청했는지 확인 후 각각 해당하는 메서드 호출
2. 요청 정보에 따라 Repository에서 BooleanBuilder를 통해 다른 조건문 생성

JeongUijeong replied:
페이지네이션 기능이 추가된다면 더불어 QueryDSL로 구현하시면 좋을 것 같습니다~

ghrltjdtprbs replied:
피드백 반영해 기존 부트캠프별 리뷰 조회 api를 삭제하고 조회에 통합했습니다! b28c0b4763136f895b5d7f2ef5e82f5247bc9393 커밋 확인해주세요!! ㅎㅎ 👍

JeongUijeong commented on src/main/java/com/fasttime/domain/review/controller/ReviewController.java:
```diff
@@ -54,4 +58,26 @@ public ResponseEntity<ResponseDTO<ReviewResponseDTO>> updateReview(
+        return ResponseEntity.ok(ResponseDTO.res(HttpStatus.OK, REVIEW_SUCCESS_MESSAGE, reviews));
+    }
+
+    @GetMapping("/by-bootcamp/summary")
```
모든 부트캠프에 대한 리뷰 요약 정보를 한 번에 조회하는 기능인 건가요? 특정 부트캠프에 대한 리뷰 요약 정보를 조회하는 API가 이 기존에 들었던 기획에 좀 더 적합한 것 같습니다!

부트캠프 목록 조회 -> 부트캠프 선택 -> 선택한 부트캠프 리뷰 요약 정보 조회 + 선택한 부트캠프 리뷰 목록 조회(페이지네이션?)

저는 이렇게 이해했었는데 혹시 제가 잘못 이해하고 있는 부분이 있다면 알려주시면 감사하겠습니다..!🥲

ghrltjdtprbs replied:
figma에 추가한 와이어프레임을 참고해주세요! 이 부분도 상의 후에 리팩토링 하는 것이 좋을 것 같습니다. api 하나 더 만들 생각도 하고 있어요 👍👍
 
<img width="306" alt="image" src="https://github.com/boocam-project/FastTime-backend/assets/105612931/df12fe54-1827-4c8f-b8eb-eaad7f41bab1">

JeongUijeong replied:
네! 감사합니다~ 이미지를 보니, 부트캠프 이름, 별점, 리뷰 개수는 목록 조회 시에 주는 게 이해가 됩니다!
하지만, 태그 총 개수, 태그별 개수는 따로 부트캠프 선택했을 때 주는 것이 좋을 것 같네요! API를 분리해야 할 것 같습니다🥲

ghrltjdtprbs replied:
리뷰 조회시 페이지네이션 적용할 것을 고려해, 부트캠프 요약 정보/ 태그 그래프 정보를 출력하는 api 를 분리하였습니다!  832d876ebcc2a785547da0311da16c2fcd58f2d4 커밋 확인해주세요! 👍👍

JeongUijeong commented on src/main/java/com/fasttime/domain/review/entity/ReviewTag.java:
```diff
@@ -28,4 +28,6 @@ public class ReviewTag {
     @JoinColumn(name = "tag_id")
     private Tag tag;
 
+    private boolean isGoodTag;
```
ReviewTag 엔터티는 Review와 Tag의 다대다 관계 연결을 위한 단순한 중간 테이블인 걸로 이해했습니다! 또한, 실제로 긍정적인지 아닌 지를 판단할 대상도 Tag 이기 때문에 Tag 테이블에 추가하는 것이 좋아 보입니다!

ghrltjdtprbs replied:
구현 중에 임의로 넣은 필드라 판단을 잘못한 것 같습니다! isGoodTag은 말씀대로 Tag 테이블에 추가하는게 좋을 것 같네요! 수정하겠습니다 ㅎㅎ 👍👍

JeongUijeong reviewed (changes requested):
수고하셨습니다~😊
몇 가지 코멘트 달아뒀으니 확인 후 답변 부탁드립니다! ㅎㅎ
특히 리뷰 조회 시 페이지네이션 기능 부재에 대해 재고해주시면 좋을 것 같아요 😊

ghrltjdtprbs committed:
[refactor] ReviewTag엔터티 isGoodTag Tag엔터티로 이동

문제 사항 :
 - isGoodTag를 ReviewTag 엔터티가 아닌 Tag 엔터티로 이동하는게 더 적절하다는 피드백

변경 사항 :
 - ReviewTag엔터티 isGoodTag Tag엔터티로 이동
 - flyway스크립트 수정
 - 관련 ReviewService 로직 수정

ghrltjdtprbs committed:
[refactor] 리뷰 전체조회 api 변경

문제 사항 :
 - getReviews api 변경 피드백

변경 사항 :
 - GET /api/v2/reviews/all -> GET /api/v2/reviews
 - 관련 SpringSecurityConfig 변경

ghrltjdtprbs committed:
[refactor] 리뷰 전체조회, 부트캠프별 조회 통합

문제 사항 :
 - 리뷰 전체조회, 부트캠프별 조회 통합 제안

변경 사항 :
 - 리뷰 전체조회, 부트캠프별 조회 통합
 - GET /api/v2/reviews로 통합
 - /api/v2/reviews/by-bootcamp 삭제

ghrltjdtprbs committed:
[refactor] 리뷰 summary, tag api 분리

문제 사항 :
 - 부트캠프 리뷰 summary와 태그 그래프 api 분리 제안

변경 사항 :
 - /api/v2/reviews/summary에 태그 응답값 제거
 - GET /api/v2/reviews/tag-graph 추가
 - 관련 로직 변경

ghrltjdtprbs committed:
[refactor] 중복 부분 함수화

문제 사항 :
 - ReviewService의 createReview,updateReview 메서드에서 중복되는 부분 함수화 제안

변경 사항 :
 - 중복 부분 updateReview 메서드로 함수화

ghrltjdtprbs committed:
[fix] 테스트 코드 수정 및 테스트케이스 추가

문제 사항 :
 - ReviewService 리팩토링으로 몇 테스트케이스 실패
 - updateReview_willSuccess, Context_getSortedReviews의 _willSuccess 메서드, Context_getBootcampReviewSummaries의 getBootcampReviewSummaries_willSuccess메서드

변경 사항 :
 - 실패하는 테스트 수정
 - getBootcampTagData() 테스트 케이스 추가

ghrltjdtprbs committed:
[fix] restdoccs 테스트코드 수정 및 추가

문제 사항 :
 - 리팩토링으로 몇 테스트케이스 실패
 - /api/v2/reviews/tag-graph api 추가

변경 사항 :
 - 실패하는 테스트 수정
 - /api/v2/reviews/tag-graph 문서화 테스트코드 작성

ghrltjdtprbs committed:
[refactor] restdoccs 문서 작성

변경 사항 :
 - 변경사항에 맞게 문서 수정
 - 새로운 api 추가

ghrltjdtprbs committed:
[refactor] ReviewService 리팩토링

변경 사항 :
 - updateReviewDetails, updateReview 메서드 합침
 - 사용하지 않는 getReviewsByBootcamp메서드 제거

ghrltjdtprbs committed:
[fix] ReviewControllerTest수정 및 테스트케이스 추가

문제 사항 :
 - 리팩토링으로 인한 테스트 코드 실패
 - 새로운 api 테스트 케이스 작성 필요

변경 사항 :
 - 실패하는 테스트 코드 수정
 - getTagCountsByBootcamp() 테스트 케이스 추가

ghrltjdtprbs committed:
[fix] ReviewServiceTest수정

문제 사항 :
 - 리팩토링으로 인한 테스트 코드 실패

변경 사항 :
 - 실패하는 테스트 코드 수정

ghrltjdtprbs committed:
[refactor] BootCampNotFoundException예외 추가

변경 사항 :
 - getSortedReviews 메서드와 getBootcampTagData 메서드에 BootCampNotFoundException 예외를 추가하였습니다.

ghrltjdtprbs committed:
[fix] ReviewServiceTest 수정

문제 사항 :
 - BootCampNotFoundException 예외 추가로 테스트 케이스 실패
변경 사항 :
 - 테스트 코드 수정 및 예외 테스트 케이스 추가

ghrltjdtprbs commented:
> 수고하셨습니다~😊 몇 가지 코멘트 달아뒀으니 확인 후 답변 부탁드립니다! ㅎㅎ 특히 리뷰 조회 시 페이지네이션 기능 부재에 대해 재고해주시면 좋을 것 같아요 😊

피드백 전부 반영해 커밋하였으니 참고해주세요! 페이지네이션은 천천히....😭

JeongUijeong reviewed (approved):
수고했어용~🩷

KwonJuHwan commented on src/main/java/com/fasttime/domain/review/controller/ReviewController.java:
```diff
@@ -54,4 +59,27 @@ public ResponseEntity<ResponseDTO<ReviewResponseDTO>> updateReview(
+        return ResponseEntity.ok(ResponseDTO.res(HttpStatus.OK, REVIEW_SUCCESS_MESSAGE, reviews));
+    }
+
+    @GetMapping("/summary")
```
해당 API 조회마다 DB에서 많은 데이터를 조회하므로, 캐싱처리가  필요해보입니다.
간편하게 Redis Caching 기능을 구현할 수 있으므로 추천드립니다.
Redis Caching :
- RedisCacheConfig를 통해, CacheManager Bean 등록
- 캐싱 처리가 필요한 서비스 로직에 @Cacheable 어노테이션 추가
- 캐싱 삭제가 필요한 서비스 로직( ex. 등록)에 @CacheEvict 어노테이션 추가
- 예시가 궁금하시면 해당 링크를 통해, 확인해주세요.
- @Cacheable 예시 - https://github.com/CoolPeace-yanolza/backend/blob/develop/src/main/java/com/coolpeace/domain/dashboard/service/DashboardService.java
- @CacheEvict 예시 - https://github.com/CoolPeace-yanolza/backend/blob/develop/src/main/java/com/coolpeace/domain/statistics/service/DailyStatisticsService.java
- RedisCacheConfig 예시 - https://github.com/CoolPeace-yanolza/backend/blob/develop/src/main/java/com/coolpeace/global/config/RedisCacheConfig.java
